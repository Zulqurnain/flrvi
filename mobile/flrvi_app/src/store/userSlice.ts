import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';
import axios from 'axios';
import { RootState } from './store';

const API_URL = 'http://localhost:5000/api/v1';

export const createProfile = createAsyncThunk<
  any,
  { birthday: any; gender: string; photos: any[] },
  { rejectValue: { msg: string } }
>('user/createProfile', async (profileData, { getState, rejectWithValue }) => {
  try {
    const { token } = (getState() as RootState).user;
    const config = {
      headers: {
        Authorization: `Bearer ${token}`,
      },
    };
    const response = await axios.post(`${API_URL}/profile`, profileData, config);
    return response.data;
  } catch (err: any) {
    return rejectWithValue(err.response.data);
  }
});

export const updateUserProfile = createAsyncThunk<
  any,
  { bio: string; photos: string[] },
  { rejectValue: { msg: string } }
>('user/updateProfile', async (profileData, { getState, rejectWithValue }) => {
  try {
    const { token } = (getState() as RootState).user;
    const config = {
      headers: {
        Authorization: `Bearer ${token}`,
      },
    };
    const response = await axios.put(`${API_URL}/profile`, profileData, config);
    return response.data;
  } catch (err: any) {
    return rejectWithValue(err.response.data);
  }
});

export const registerUser = createAsyncThunk<
  { token: string },
  { name: string; email: string; password: string },
  { rejectValue: { msg: string } }
>('user/register', async (userData, { rejectWithValue }) => {
  try {
    const response = await axios.post(`${API_URL}/users/register`, userData);
    return response.data;
  } catch (err: any) {
    return rejectWithValue(err.response.data);
  }
});

export const loginUser = createAsyncThunk<
  { token: string },
  { email: string; password: string },
  { rejectValue: { msg: string } }
>('user/login', async (credentials, { rejectWithValue }) => {
  try {
    const response = await axios.post(`${API_URL}/users/login`, credentials);
    return response.data;
  } catch (err: any) {
    return rejectWithValue(err.response.data);
  }
});

export const fetchProfiles = createAsyncThunk<
  any[],
  void,
  { rejectValue: { msg: string } }
>('user/fetchProfiles', async (_, { rejectWithValue }) => {
  try {
    const response = await axios.get(`${API_URL}/profile/search`);
    return response.data;
  } catch (err: any) {
    return rejectWithValue(err.response.data);
  }
});

export const fetchConversations = createAsyncThunk<
  any[],
  void,
  { rejectValue: { msg: string } }
>('user/fetchConversations', async (_, { rejectWithValue }) => {
  try {
    const response = await axios.get(`${API_URL}/chat/conversations`);
    return response.data;
  } catch (err: any) {
    return rejectWithValue(err.response.data);
  }
});

export const fetchUserProfile = createAsyncThunk<
  any,
  string,
  { rejectValue: { msg: string } }
>('user/fetchUserProfile', async (userId, { rejectWithValue }) => {
  try {
    const response = await axios.get(`${API_URL}/profile/user/${userId}`);
    return response.data;
  } catch (err: any) {
    return rejectWithValue(err.response.data);
  }
});

export const fetchCurrentUserProfile = createAsyncThunk<
  any,
  void,
  { rejectValue: { msg: string } }
>('user/fetchCurrentUserProfile', async (_, { getState, rejectWithValue }) => {
  try {
    const { token } = (getState() as RootState).user;
    const config = {
      headers: {
        Authorization: `Bearer ${token}`,
      },
    };
    const response = await axios.get(`${API_URL}/profile/me`, config);
    return response.data;
  } catch (err: any) {
    return rejectWithValue(err.response.data);
  }
});

export const likeUserProfile = createAsyncThunk<
  any,
  string,
  { rejectValue: { msg: string } }
>('user/likeUserProfile', async (userId, { rejectWithValue }) => {
  try {
    const response = await axios.post(`${API_URL}/profile/like/${userId}`);
    return response.data;
  } catch (err: any) {
    return rejectWithValue(err.response.data);
  }
});

export const subscribeUser = createAsyncThunk<
  any,
  { email: string; plan: string; source: string },
  { rejectValue: { msg: string } }
>('user/subscribe', async (subscriptionData, { getState, rejectWithValue }) => {
  try {
    const { token } = (getState() as RootState).user;
    const config = {
      headers: {
        Authorization: `Bearer ${token}`,
      },
    };
    const response = await axios.post(`${API_URL}/payment/subscribe`, subscriptionData, config);
    return response.data;
  } catch (err: any) {
    return rejectWithValue(err.response.data);
  }
});

export const purchaseBoost = createAsyncThunk<
  any,
  { productId: string; amount: number },
  { rejectValue: { msg: string } }
>('user/purchaseBoost', async (purchaseData, { getState, rejectWithValue }) => {
  try {
    const { token } = (getState() as RootState).user;
    const config = {
      headers: {
        Authorization: `Bearer ${token}`,
      },
    };
    const response = await axios.post(`${API_URL}/payment/purchase-consumable`, purchaseData, config);
    return response.data;
  } catch (err: any) {
    return rejectWithValue(err.response.data);
  }
});

export const fetchBlockedUsers = createAsyncThunk<
  any[],
  void,
  { rejectValue: { msg: string } }
>('user/fetchBlockedUsers', async (_, { getState, rejectWithValue }) => {
  try {
    const { token } = (getState() as RootState).user;
    const config = {
      headers: {
        Authorization: `Bearer ${token}`,
      },
    };
    const response = await axios.get(`${API_URL}/users/blocked`, config);
    return response.data;
  } catch (err: any) {
    return rejectWithValue(err.response.data);
  }
});

export const unblockUser = createAsyncThunk<
  any,
  string,
  { rejectValue: { msg: string } }
>('user/unblockUser', async (userId, { getState, rejectWithValue }) => {
  try {
    const { token } = (getState() as RootState).user;
    const config = {
      headers: {
        Authorization: `Bearer ${token}`,
      },
    };
    const response = await axios.post(`${API_URL}/users/unblock`, { userId }, config);
    return response.data;
  } catch (err: any) {
    return rejectWithValue(err.response.data);
  }
});

export const updateNotificationSetting = createAsyncThunk<
  any,
  { [key: string]: boolean },
  { rejectValue: { msg: string } }
>('user/updateNotificationSetting', async (settings, { getState, rejectWithValue }) => {
  try {
    const { token } = (getState() as RootState).user;
    const config = {
      headers: {
        Authorization: `Bearer ${token}`,
      },
    };
    const response = await axios.post(`${API_URL}/users/settings/notifications`, settings, config);
    return response.data;
  } catch (err: any) {
    return rejectWithValue(err.response.data);
  }
});

export const setLanguage = createAsyncThunk<
  any,
  string,
  { rejectValue: { msg: string } }
>('user/setLanguage', async (languageCode, { getState, rejectWithValue }) => {
  try {
    const { token } = (getState() as RootState).user;
    const config = {
      headers: {
        Authorization: `Bearer ${token}`,
      },
    };
    const response = await axios.post(`${API_URL}/users/settings/language`, { language: languageCode }, config);
    return response.data;
  } catch (err: any) {
    return rejectWithValue(err.response.data);
  }
});

export const fetchTransactionHistory = createAsyncThunk<
  any[],
  void,
  { rejectValue: { msg: string } }
>('user/fetchTransactionHistory', async (_, { getState, rejectWithValue }) => {
  try {
    const { token } = (getState() as RootState).user;
    const config = {
      headers: {
        Authorization: `Bearer ${token}`,
      },
    };
    const response = await axios.get(`${API_URL}/payment/history`, config);
    return response.data;
  } catch (err: any) {
    return rejectWithValue(err.response.data);
  }
});

const userSlice = createSlice({
  name: 'user',
  initialState: {
    token: null,
    profile: null,
    profiles: [],
    conversations: [],
    blockedUsers: [],
    transactions: [],
    status: 'idle',
    error: null as string | null,
  },
  reducers: {
    logout: (state) => {
      state.token = null;
    },
  },
  extraReducers: (builder) => {
    builder
      .addCase(loginUser.pending, (state) => {
        state.status = 'loading';
      })
      .addCase(loginUser.fulfilled, (state, action) => {
        state.status = 'succeeded';
        state.token = action.payload.token;
      })
      .addCase(loginUser.rejected, (state, action) => {
        state.status = 'failed';
        state.error = action.payload ? (action.payload as { msg: string }).msg : 'Login failed';
      })
      .addCase(registerUser.pending, (state) => {
        state.status = 'loading';
      })
      .addCase(registerUser.fulfilled, (state, action) => {
        state.status = 'succeeded';
        state.token = action.payload.token;
      })
      .addCase(registerUser.rejected, (state, action) => {
        state.status = 'failed';
        state.error = action.payload ? (action.payload as { msg: string }).msg : 'Registration failed';
      })
      .addCase(createProfile.pending, (state) => {
        state.status = 'loading';
      })
      .addCase(createProfile.fulfilled, (state, action) => {
        state.status = 'succeeded';
        state.profile = action.payload;
      })
      .addCase(createProfile.rejected, (state, action) => {
        state.status = 'failed';
        state.error = action.payload ? (action.payload as { msg: string }).msg : 'Profile creation failed';
      })
      .addCase(updateUserProfile.pending, (state) => {
        state.status = 'loading';
      })
      .addCase(updateUserProfile.fulfilled, (state, action) => {
        state.status = 'succeeded';
        state.profile = action.payload;
      })
      .addCase(updateUserProfile.rejected, (state, action) => {
        state.status = 'failed';
        state.error = action.payload ? (action.payload as { msg: string }).msg : 'Profile update failed';
      })
      .addCase(fetchProfiles.pending, (state) => {
        state.status = 'loading';
      })
      .addCase(fetchProfiles.fulfilled, (state, action) => {
        state.status = 'succeeded';
        state.profiles = action.payload;
      })
      .addCase(fetchProfiles.rejected, (state, action) => {
        state.status = 'failed';
        state.error = action.payload ? (action.payload as { msg: string }).msg : 'Fetching profiles failed';
      })
      .addCase(fetchConversations.pending, (state) => {
        state.status = 'loading';
      })
      .addCase(fetchConversations.fulfilled, (state, action) => {
        state.status = 'succeeded';
        state.conversations = action.payload;
      })
      .addCase(fetchConversations.rejected, (state, action) => {
        state.status = 'failed';
        state.error = action.payload ? (action.payload as { msg: string }).msg : 'Fetching conversations failed';
      })
      .addCase(fetchUserProfile.pending, (state) => {
        state.status = 'loading';
      })
      .addCase(fetchUserProfile.fulfilled, (state, action) => {
        state.status = 'succeeded';
        state.profile = action.payload;
      })
      .addCase(fetchUserProfile.rejected, (state, action) => {
        state.status = 'failed';
        state.error = action.payload ? (action.payload as { msg: string }).msg : 'Fetching user profile failed';
      })
      .addCase(likeUserProfile.pending, (state) => {
        state.status = 'loading';
      })
      .addCase(likeUserProfile.fulfilled, (state, action) => {
        state.status = 'succeeded';
      })
      .addCase(likeUserProfile.rejected, (state, action) => {
        state.status = 'failed';
        state.error = action.payload ? (action.payload as { msg: string }).msg : 'Liking user profile failed';
      })
      .addCase(fetchCurrentUserProfile.pending, (state) => {
        state.status = 'loading';
      })
      .addCase(fetchCurrentUserProfile.fulfilled, (state, action) => {
        state.status = 'succeeded';
        state.profile = action.payload;
      })
      .addCase(fetchCurrentUserProfile.rejected, (state, action) => {
        state.status = 'failed';
        state.error = action.payload ? (action.payload as { msg: string }).msg : 'Fetching current user profile failed';
      })
      .addCase(subscribeUser.pending, (state) => {
        state.status = 'loading';
      })
      .addCase(subscribeUser.fulfilled, (state, action) => {
        state.status = 'succeeded';
        // Update user profile to reflect premium status
        if (state.profile) {
          state.profile.isPremium = true;
        }
      })
      .addCase(subscribeUser.rejected, (state, action) => {
        state.status = 'failed';
        state.error = action.payload ? (action.payload as { msg: string }).msg : 'Subscription failed';
      })
      .addCase(purchaseBoost.pending, (state) => {
        state.status = 'loading';
      })
      .addCase(purchaseBoost.fulfilled, (state, action) => {
        state.status = 'succeeded';
        // Update user profile to reflect boost status if needed
        // This could involve setting a boost expiration timestamp
      })
      .addCase(purchaseBoost.rejected, (state, action) => {
        state.status = 'failed';
        state.error = action.payload ? (action.payload as { msg: string }).msg : 'Boost purchase failed';
      })
      .addCase(fetchBlockedUsers.pending, (state) => {
        state.status = 'loading';
      })
      .addCase(fetchBlockedUsers.fulfilled, (state, action) => {
        state.status = 'succeeded';
        state.blockedUsers = action.payload;
      })
      .addCase(fetchBlockedUsers.rejected, (state, action) => {
        state.status = 'failed';
        state.error = action.payload ? (action.payload as { msg: string }).msg : 'Failed to fetch blocked users';
      })
      .addCase(unblockUser.pending, (state) => {
        state.status = 'loading';
      })
      .addCase(unblockUser.fulfilled, (state, action) => {
        state.status = 'succeeded';
        // Remove the unblocked user from the blocked users list
        if (state.blockedUsers) {
          state.blockedUsers = state.blockedUsers.filter((user: any) => user.id !== action.meta.arg);
        }
      })
      .addCase(unblockUser.rejected, (state, action) => {
        state.status = 'failed';
        state.error = action.payload ? (action.payload as { msg: string }).msg : 'Failed to unblock user';
      })
      .addCase(updateNotificationSetting.pending, (state) => {
        state.status = 'loading';
      })
      .addCase(updateNotificationSetting.fulfilled, (state, action) => {
        state.status = 'succeeded';
        // Update the user's notification settings
        if (state.profile) {
          state.profile.notificationSettings = {
            ...state.profile.notificationSettings,
            ...action.meta.arg
          };
        }
      })
      .addCase(updateNotificationSetting.rejected, (state, action) => {
        state.status = 'failed';
        state.error = action.payload ? (action.payload as { msg: string }).msg : 'Failed to update notification settings';
      })
      .addCase(setLanguage.pending, (state) => {
        state.status = 'loading';
      })
      .addCase(setLanguage.fulfilled, (state, action) => {
        state.status = 'succeeded';
        // Update the user's language preference
        if (state.profile) {
          state.profile.language = action.meta.arg;
        }
      })
      .addCase(setLanguage.rejected, (state, action) => {
        state.status = 'failed';
        state.error = action.payload ? (action.payload as { msg: string }).msg : 'Failed to set language';
      })
      .addCase(fetchTransactionHistory.pending, (state) => {
        state.status = 'loading';
      })
      .addCase(fetchTransactionHistory.fulfilled, (state, action) => {
        state.status = 'succeeded';
        state.transactions = action.payload;
      })
      .addCase(fetchTransactionHistory.rejected, (state, action) => {
        state.status = 'failed';
        state.error = action.payload ? (action.payload as { msg: string }).msg : 'Failed to fetch transaction history';
      });
  },
});

export const { logout } = userSlice.actions;
export const selectUserRegistrationStatus = (state: RootState) => state.user.status;
export const selectSubscriptionStatus = (state: RootState) => state.user.profile?.isPremium || false;
export const selectBoostPurchaseStatus = (state: RootState) => state.user.status;
export const selectBlockedUsers = (state: RootState) => state.user.blockedUsers || [];
export const selectNotificationSettings = (state: RootState) => state.user.profile?.notificationSettings || {};
export const selectCurrentLanguage = (state: RootState) => state.user.profile?.language || 'en';
export const selectAvailableLanguages = () => [
  { code: 'en', name: 'English' },
  { code: 'th', name: 'Thai' }
];
export const selectTransactionHistory = (state: RootState) => state.user.transactions || [];
export default userSlice.reducer;
