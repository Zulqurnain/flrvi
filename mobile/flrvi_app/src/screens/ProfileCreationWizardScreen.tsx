import React, { useState } from 'react';
import { View, Text, StyleSheet, Button, Alert, Image, TouchableOpacity } from 'react-native';
import { useDispatch } from 'react-redux';
import { createProfile } from '../store/userSlice';
import { AppDispatch } from '../store/store';
import DatePicker from 'react-native-date-picker';
import SegmentedControl from '@react-native-segmented-control/segmented-control';
import { launchImageLibrary } from 'react-native-image-picker';

const ProfileCreationWizardScreen = ({ navigation }) => {
  const [step, setStep] = useState(1);
  const [birthday, setBirthday] = useState(new Date());
  const [gender, setGender] = useState('Man');
  const [photos, setPhotos] = useState<string[]>([]);
  const dispatch = useDispatch<AppDispatch>();

  const handleNext = () => {
    if (step === 3) {
      if (photos.length === 0) {
        Alert.alert('Validation Error', 'Please upload at least one photo.');
        return;
      }
      const profileData = { birthday, gender, photos };
      dispatch(createProfile(profileData)).then(() => {
        navigation.navigate('HomeScreen');
      });
    } else {
      setStep(step + 1);
    }
  };

  const handlePhotoUpload = () => {
    launchImageLibrary({ mediaType: 'photo', selectionLimit: 6 }, (response) => {
      if (response.didCancel) {
        console.log('User cancelled image picker');
      } else if (response.errorCode) {
        console.log('ImagePicker Error: ', response.errorMessage);
      } else if (response.assets) {
        const uris = response.assets.map((asset) => asset.uri).filter((uri): uri is string => !!uri);
        setPhotos(uris);
      }
    });
  };

  return (
    <View style={styles.container}>
      {step === 1 && (
        <View style={styles.stepContainer}>
          <Text style={styles.title}>Basic Info</Text>
          <Text style={styles.label}>Your Birthday</Text>
          <DatePicker date={birthday} onDateChange={setBirthday} mode="date" />
          <Text style={styles.label}>Your Gender</Text>
          <SegmentedControl
            values={['Man', 'Woman', 'Other']}
            selectedIndex={['Man', 'Woman', 'Other'].indexOf(gender)}
            onChange={(event) => {
              setGender(event.nativeEvent.value);
            }}
          />
        </View>
      )}
      {step === 2 && (
        <View style={styles.stepContainer}>
          <Text style={styles.title}>Add Photos</Text>
          <Button title="Upload Photos" onPress={handlePhotoUpload} />
          <View style={styles.photoGrid}>
            {photos.map((uri, index) => (
              <Image key={index} source={{ uri }} style={styles.thumbnail} />
            ))}
          </View>
        </View>
      )}
      {step === 3 && (
        <View style={styles.stepContainer}>
          <Text style={styles.title}>Your Location</Text>
          <Button title="Enable Location Services" onPress={() => Alert.alert('Location', 'Location services enabled.')} />
        </View>
      )}
      <Button title={step === 3 ? 'Finish' : 'Next'} onPress={handleNext} />
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'space-between',
    padding: 24,
    backgroundColor: '#F5F5F5',
  },
  stepContainer: {
    flex: 1,
    justifyContent: 'center',
  },
  title: {
    fontSize: 24,
    fontWeight: 'bold',
    marginBottom: 40,
    color: '#333333',
    textAlign: 'center',
  },
  label: {
    fontSize: 16,
    fontWeight: '600',
    color: '#333333',
    marginBottom: 10,
    marginTop: 20,
  },
  photoGrid: {
    flexDirection: 'row',
    flexWrap: 'wrap',
    marginTop: 20,
  },
  thumbnail: {
    width: 100,
    height: 100,
    margin: 5,
  },
});

export default ProfileCreationWizardScreen;
