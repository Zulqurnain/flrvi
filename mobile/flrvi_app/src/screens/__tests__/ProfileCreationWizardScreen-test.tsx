import React from 'react';
import renderer from 'react-test-renderer';
import ProfileCreationWizardScreen from '../ProfileCreationWizardScreen';
import { Provider } from 'react-redux';
import { store } from '../../store/store';

it('renders correctly', () => {
  const navigation = { navigate: jest.fn() };
  const tree = renderer.create(
    <Provider store={store}>
      <ProfileCreationWizardScreen navigation={navigation} />
    </Provider>
  ).toJSON();
  expect(tree).toMatchSnapshot();
});
