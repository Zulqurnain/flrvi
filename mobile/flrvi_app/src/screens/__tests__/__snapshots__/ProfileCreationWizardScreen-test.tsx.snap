// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    {
      "backgroundColor": "#F5F5F5",
      "flex": 1,
      "justifyContent": "space-between",
      "padding": 24,
    }
  }
>
  <View
    style={
      {
        "flex": 1,
        "justifyContent": "center",
      }
    }
  >
    <Text
      style={
        {
          "color": "#333333",
          "fontSize": 24,
          "fontWeight": "bold",
          "marginBottom": 40,
          "textAlign": "center",
        }
      }
    >
      Basic Info
    </Text>
    <Text
      style={
        {
          "color": "#333333",
          "fontSize": 16,
          "fontWeight": "600",
          "marginBottom": 10,
          "marginTop": 20,
        }
      }
    >
      Your Birthday
    </Text>
    <RNDatePicker
      buttonColor="#000000"
      cancelText="Cancel"
      confirmText="Confirm"
      date="2025-08-09T19:26:04.188Z"
      dividerColor="#000000"
      minuteInterval={1}
      mode="date"
      onChange={[Function]}
      onDateChange={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        [
          {
            "height": 216,
            "width": 310,
          },
          undefined,
        ]
      }
      textColor="#000000"
      theme="light"
      timeZoneOffsetInMinutes=""
      title="Select date"
    />
    <Text
      style={
        {
          "color": "#333333",
          "fontSize": 16,
          "fontWeight": "600",
          "marginBottom": 10,
          "marginTop": 20,
        }
      }
    >
      Your Gender
    </Text>
    <RNCSegmentedControl
      enabled={true}
      onChange={[Function]}
      selectedIndex={0}
      style={
        [
          {
            "height": 32,
          },
          undefined,
        ]
      }
      values={
        [
          "Man",
          "Woman",
          "Other",
        ]
      }
    />
  </View>
  <View
    accessibilityRole="button"
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        [
          {},
        ]
      }
    >
      <Text
        style={
          [
            {
              "color": "#007AFF",
              "fontSize": 18,
              "margin": 8,
              "textAlign": "center",
            },
          ]
        }
      >
        Next
      </Text>
    </View>
  </View>
</View>
`;
